src/persist-client/src/batch.rs

```diff
@@ -307,12 +264,234 @@ where
         }
         self.records.push(((&self.key_buf, &self.val_buf), t, d));
 
-        // TODO(aljoscha): As of now, this batch builder doesn't have constant
-        // memory usage but scales with the number of added records. Instead of
-        // keeping one records builder around, we should periodically finish and
-        // flush the updates to a blob, store the blob key in the builder, and
-        // reset the records builder.
+        // If we've filled up a chunk of ColumnarRecords, flush it out now to
+        // blob storage to keep our memory usage capped.
+        if let Some(parts) = self.records.take_filled() {
+            for part in parts {
+                // TODO: This upper would ideally be `[self.max_ts+1]` but
+                // there's nothing that lets us increment a timestamp. An empty
+                // antichain is guaranteed to correctly bound the data in this
+                // part, but it doesn't really tell us anything. Figure out how
+                // to make a tighter bound, possibly by changing the part
+                // description to be an _inclusive_ upper.
```

Comment:
Yeah, I've dabbled with this _inclusive upper_ in my persistent source examples. But it can be confusing to people that are used to timely/dd uppers. No need to change anything here, btw!

Also, nice that you thought of the empty antichain as a valid upper to get around this problem for now!